R/plot.R

```diff
@@ -0,0 +1,449 @@
+
+#' A histogram of density ratio estimates
+#'
+#' Creates a histogram of the density ratio estimates. Useful to understand the
+#' distribution of estimated density ratios in each sample, or compare it among
+#' samples. It is the default plotting method for density ratio objects.
+#'
+#' @param object Density ratio object created with e.g., [kliep()], [ulsif()],
+#'   or [naive()]
+#' @param sample Character string indicating whether to plot the 'numerator',
+#'   'denominator', or 'both' samples. Default is 'both'.
+#' @param logscale Logical indicating whether to plot the density ratio
+#'   estimates on a log scale. Default is TRUE.
+#' @param binwidth Numeric indicating the width of the bins, passed on to
+#'   `ggplot2`.
+#' @param bins Numeric indicating the number of bins. Overriden by binwidth, and
+#'   passed on to `ggplot2`.
+#' @param ... Additional arguments passed on to `predict()`.
+#'
+#' @return A histogram of density ratio estimates.
+#'
+#'
+#' @examples
+dr.histogram <- function(object,
+                         samples = "both",
+                         logscale = TRUE,
+                         binwidth = NULL,
+                         bins = NULL,
+                         ...) {
+
+  # Check object type
+  check.object.type(object)
+
+
+  # Create data object and estimate density ratio
+  data <- rbind(object$df_numerator, object$df_denominator)
+  # Check no variable names that will be overriden when plotting.
+  check.overriden.names(data)
+  # Create density ratio prediction
+  data$dr <- predict(object, newdata = data, ...)
+
+
+  if (logscale) {
+
+    # Convert negative predicted density ratios to 10e-3, so log can be computed
+    if(any(data$dr <= 0)){
+      data$dr[data$dr <= 0] <- 10e-3
+      # Throw warning with number of converted values
+      count <- length(data$dr[data$dr <= 0])
+      warning(
+        paste("Negative estimated density ratios for", count, "observation(s) converted to 10e-3 before applying logarithmic transformation"),
+        call. = FALSE)
+    }
+
+    # Apply log transformation
+    data$dr <- log(data$dr)
+    x_lab <- "Log (Density Ratio)"
+  } else {
+    x_lab <- "Density Ratio"
+  }
+
+  # Create a sample index variable (denominator or numerator)
+  data$sample <- rep(c("numerator", "denominator"),
+                     c(nrow(object$df_numerator), nrow(object$df_denominator)))
+
+  # Create a object selection variable (both, numerator, denominator)
+  obsselect <- match.arg(samples, c("both", "numerator", "denominator"))
+
+  # If not both, subset data (only num or only den)
+  if (obsselect != "both") {
+    data <- filter(data, sample == obsselect)
+  }
+
+  # Plot
+  plot <-
+    ggplot(data, aes(x = dr)) +
```

Comment:
Then here `x = ext$dr`

Reply:
Yes or just ggplot(ext, aes(x = dr))

Reply:
Also possible indeed, but I think I like being consistent (so it is always clear and transparent where `dr` comes from).

Reply:
ok!